packages/flutter/lib/src/widgets/undo_history.dart

```diff
@@ -0,0 +1,466 @@
+  });
+
+  /// The value to track over time.
+  final ValueNotifier<T> value;
+
+  /// Called when checking whether a value change should be pushed onto
+  /// the undo stack.
+  final bool Function(T? oldValue, T newValue)? shouldChangeUndoStack;
+
+  /// Called when an undo or redo causes a state change.
+  ///
+  /// If the state would still be the same before and after the undo/redo, this
+  /// will not be called. For example, receiving a redo when there is nothing
+  /// to redo will not call this method.
+  final void Function(T value) onTriggered;
+
+  /// The [FocusNode] that will be used to listen for focus to set the initial
+  /// undo state for the element.
+  final FocusNode focusNode;
+
+  /// {@template flutter.widgets.undoHistory.controller}
+  /// Controls the undo state.
+  ///
+  /// If null, this widget will create its own [UndoHistoryController].
+  /// {@endtemplate}
+  final UndoHistoryController? controller;
+
+  /// The child widget of [UndoHistory].
+  final Widget child;
+
+  @override
+  State<UndoHistory<T>> createState() => _UndoHistoryState<T>();
+}
+
+/// State for a [UndoHistory].
+///
+/// Provides [undo], [redo], [canUndo], and [canRedo] for programmatic access
+/// to the undo state for custom undo and redo UI implementations.
+class _UndoHistoryState<T> extends State<UndoHistory<T>> with UndoManagerClient {
+  final _UndoStack<T> _stack = _UndoStack<T>();
+  late final _Throttled<T> _throttledPush;
+  Timer? _throttleTimer;
+
+  // This duration was chosen as a best fit for the behavior of Mac, Linux,
+  // and Windows undo/redo state save durations, but it is not perfect for any
+  // of them.
+  static const Duration _kThrottleDuration = Duration(milliseconds: 500);
+
+  // Record the last value to prevent pushing multiple
+  // of the same value in a row onto the undo stack. For example, _push gets
+  // called both in initState and when the EditableText receives focus.
+  T? _lastValue;
+
+  UndoHistoryController? _controller;
+
+  UndoHistoryController get _effectiveController => widget.controller ?? (_controller ??= UndoHistoryController());
+
+  @override
+  void undo() {
+    _update(_stack.undo());
+    _updateState();
+  }
+
+  @override
+  void redo() {
+    _update(_stack.redo());
+    _updateState();
+  }
+
+  @override
+  bool get canUndo => _stack.canUndo;
+
+  @override
+  bool get canRedo => _stack.canRedo;
+
+  void _updateState() {
+    _effectiveController.value = UndoHistoryValue(canUndo: canUndo, canRedo: canRedo);
+
+    if (defaultTargetPlatform != TargetPlatform.iOS) {
+      return;
+    }
+
+    if (UndoManager.client == this) {
+      UndoManager.setUndoState(canUndo: canUndo, canRedo: canRedo);
+    }
+  }
+
+  void _undoFromIntent(UndoTextIntent intent) {
+    undo();
+  }
+
+  void _redoFromIntent(RedoTextIntent intent) {
+    redo();
+  }
+
+  void _update(T? nextValue) {
+    if (nextValue == null) {
+      return;
+    }
+    if (nextValue == _lastValue) {
```

Comment:
Will this also update if only selection changes?

Reply:
The logic of whether to update the undo stack or not is controlled by the implementation of `shouldChangeUndoStack` on `UndoHistory` – for `EditableText`, no, only text changes will push onto the stack: https://github.com/flutter/flutter/pull/98294/files#diff-f5f93c879cef9a102adbf148583de9f3d9a05f9678fdcad7bde106f62200bb34R3722